### PR TITLE
EC-341 Explorer - add ActionBuilder

### PR DESCRIPTION
**Pull request by GusevTimofey**

### CODE REVIEW

File: conf/routes [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
 GET        /api/history/bestHeaderAt/:height                             controllers.HistoryController.findBestHeaderAtHeightApi(height: Int)
 GET        /history/bestHeaderAt/:height                                 controllers.HistoryController.findBestHeaderAtHeightView(height: Int)
 GET        /api/history/lastHeaders/:qty                                 controllers.HistoryController.listLastHeadersApi(qty: Int)
 GET        /history/lastHeaders/:qty                                     controllers.HistoryController.listLastHeadersView(qty: Int)
 GET        /api/history/headers/range/:from/:to                          controllers.HistoryController.listHeadersByHeightRangeApi(from: Int, to: Int)
 GET        /history/headers/range/:from/:to                              controllers.HistoryController.listHeadersByHeightRangeView(from: Int, to: Int)
+GET        /findHeadersByDate/:date/:count                               controllers.HistoryController.findHeadersByDateView(date: String, count: Int)
+GET        /api/findHeadersByDate/:date/:count                           controllers.HistoryController.findHeadersByDateApi(date: String, count: Int)
```
Comment (jhnsmth) [inline conf/routes]: this route should have a 'history' prefix

File: app/utils/CheckAction.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  override def invokeBlock[A](request: Request[A], block: Request[A] => Future[Result]): Future[Result] =
+    if (from >= 0 && count > 0) block(request) else resolve(Results.BadRequest)
+}
+
+class FromCountCheckActionFactory @Inject()(parser: BodyParsers.Default) {
+  def apply(from: Int, count: Int): FromToCheckAction = new FromToCheckAction(parser, from, count)
+}
+
+class DateFromCountAction(parser: BodyParsers.Default, date: String, count: Int) extends ActionBuilderImpl(parser) {
+  private val sdf: SimpleDateFormat = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss")
```
Comment (jhnsmth) [inline app/utils/CheckAction.scala]: Add `.setLenient(false)` to make date validations more strict, so inputs like `2018-07-40` won't be valid

File: app/utils/CheckAction.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import scala.concurrent.Future
+import scala.concurrent.Future.{successful => resolve}
+import scala.util.Try
+
+class Base16CheckAction(parser: BodyParsers.Default, modifierId: String) extends ActionBuilderImpl(parser) {
+  override def invokeBlock[A](request: Request[A], block: Request[A] => Future[Result]): Future[Result] =
+    if (Base16.decode(modifierId).isSuccess) block(request) else resolve(Results.Forbidden)
+}
+
+class Base16CheckActionFactory @Inject()(parser: BodyParsers.Default) {
```
Comment (GRYE) [inline app/utils/CheckAction.scala]: can't you use companion object here?

File: app/controllers/HistoryController.scala [hunk 4]
```diff
@@ -1,131 +1,194 @@
...
-      case list: List[Header] => Ok(getHeaderListView(list))
-    }
+  def listHeadersByCountApi(from: Int, count: Int): Action[AnyContent] = fromCountCheck(from, count).async {
+    historyDao
+      .listHeadersByCount(from, count)
+      .map {
+        case Nil => NotFound
+        case list: List[Header] => Ok(list.asJson)
+      }
+      .recover {
```
Comment (GRYE) [inline app/controllers/HistoryController.scala]: You can actually move recovering to action definition

File: app/modules/MainModule.scala [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import settings.ExplorerSettings._
+
+class MainModule extends AbstractModule with ScalaModule {
+
+  @Provides
+  def provideExplorerAppSettings(configuration: Configuration): ExplorerSettings = {
+
+    val settings: ExplorerSettings = ConfigFactory.load("local.conf")
+      .withFallback(configuration.underlying).as[ExplorerSettings]("encry")
+
```
Comment (GusevTimofey) [inline app/modules/MainModule.scala]: fix

File: app/settings/ExplorerSettings.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import com.typesafe.config.Config
+import net.ceedubs.ficus.readers.ValueReader
+
+case class ExplorerSettings(postgres: PostgresSettings)
+
+object ExplorerSettings {
+
+  implicit val fileReader: ValueReader[File] = (cfg, path) => new File(cfg.getString(path))
+  implicit val byteValueReader: ValueReader[Byte] = (cfg, path) => cfg.getInt(path).toByte
+  implicit val inetSocketAddressReader: ValueReader[InetSocketAddress] = { (config: Config, path: String) =>
```
Comment (GusevTimofey) [inline app/settings/ExplorerSettings.scala]: ?

File: app/utils/CheckAction.scala [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+class HeightCheckAction(parser: BodyParsers.Default, height: Int) extends ActionBuilderImpl(parser) {
+  override def invokeBlock[A](request: Request[A], block: Request[A] => Future[Result]): Future[Result] =
+    if (height >= 0)
+      block(request).recoverWith {
+        case NonFatal(_) => resolve(Results.BadRequest)
+      }(executionContext)
+    else
+      resolve(Results.BadRequest)
```
Comment (kapinuss) [inline app/utils/CheckAction.scala]: do it on the same string

**Pull request merged**